pkg/utils/opsqueue.go

```diff
@@ -57,42 +58,67 @@ func (oq *OpsQueue) Start() {
 	go oq.process()
 }
 
-func (oq *OpsQueue) Stop() {
+func (oq *OpsQueue) Stop() <-chan struct{} {
 	oq.lock.Lock()
 	if oq.isStopped {
 		oq.lock.Unlock()
-		return
+		return oq.doneChan
 	}
 
 	oq.isStopped = true
-	close(oq.ops)
+	close(oq.wake)
 	oq.lock.Unlock()
-}
-
-func (oq *OpsQueue) IsStarted() bool {
-	oq.lock.RLock()
-	defer oq.lock.RUnlock()
-
-	return oq.isStarted
+	return oq.doneChan
 }
 
 func (oq *OpsQueue) Enqueue(op func()) {
-	oq.lock.RLock()
-	if oq.isStopped {
-		oq.lock.RUnlock()
-		return
-	}
-
-	select {
-	case oq.ops <- op:
-	default:
-		oq.logger.Errorw("ops queue full", nil, "name", oq.name, "size", oq.size)
+	oq.lock.Lock()
+	defer oq.lock.Unlock()
+
+	oq.ops.PushBack(op)
+	if oq.ops.Len() == 1 && !oq.isStopped {
+		select {
+		case oq.wake <- struct{}{}:
+		default:
+		}
 	}
-	oq.lock.RUnlock()
 }
 
 func (oq *OpsQueue) process() {
-	for op := range oq.ops {
-		op()
+done:
+	for {
+		<-oq.wake
+		for {
+			oq.lock.Lock()
+			if oq.isStopped {
+				oq.lock.Unlock()
+				break done
+			}
+
+			if oq.ops.Len() == 0 {
+				oq.lock.Unlock()
+				break
+			}
+			op := oq.ops.PopFront()
+			oq.lock.Unlock()
+
+			op()
+		}
 	}
+
+	if oq.flushOnStop {
+		for {
+			oq.lock.Lock()
+			if oq.ops.Len() == 0 {
+				oq.lock.Unlock()
+				break
+			}
+			op := oq.ops.PopFront()
+			oq.lock.Unlock()
+
+			op()
+		}
+	}
```

Comment:
could this be removed if the condition above was `oq.IsStopped && !oq.flushOnStop`?

Reply:
Yeah, this  is ugly, but could not figure out a way. If the condition above is changed, for queues where `flushOnStop: true`, even after stop, there is no way to exit the for loop :-(. Let me think a bit more and see if we can get rid of this.

Reply:
Can make it like
```
done:                                                                                                                                  
      for {                                                                                                                              
          <-oq.wake                                                                                                                      
          for {                                                                                                                          
              oq.lock.Lock()                                                                                                             
~             if oq.isStopped && !oq.flushOnStop {                                                                                       
                  oq.lock.Unlock()                                                                                                       
                  break done                                                                                                             
              }                                                                                                                          
                                                                                                                                         
              if oq.ops.Len() == 0 {                                                                                                     
                  oq.lock.Unlock()                                                                                                       
~                 if oq.isStopped {                                                                                                      
+                     break done                                                                                                         
+                 } else {                                                                                                               
+                     break                                                                                                              
+                 }                                                                                                                      
              }                                                                                                                          
              op := oq.ops.PopFront()                                                                                                    
              oq.lock.Unlock()                                                                                                           
                                                                                                                                         
              op()                                                                                                                       
          }                                                                                                                              
      } 
```
`break done` vs `break` based on an `if` is not great, but at least avoids repeated code. WDYT?

Reply:
it would have to be `oq.isStopped && !oq.flushOnStop || oq.ops.Len() == 0`... could also use return and defer the channel close to get rid of the loop label

Reply:
One more check please @paulwe https://github.com/livekit/livekit/pull/2418/commits/a7db028484ac6122d251a7c758a3018633830bcf

Reply:
lgtm!

Reply:
Thank you 🙏